graphics/pl240src/man/man1/pl.1
.ig >>
<STYLE TYPE="text/css">
<!--
        A:link{text-decoration:none}
        A:visited{text-decoration:none}
        A:active{text-decoration:none}
        OL,UL,P,BODY,TD,TR,TH,FORM { font-family: arial,helvetica,sans-serif;; font-size:small; color: #333333; }

        H1 { font-size: x-large; font-family: arial,helvetica,sans-serif; }
        H2 { font-size: large; font-family: arial,helvetica,sans-serif; }
        H3 { font-size: medium; font-family: arial,helvetica,sans-serif; }
        H4 { font-size: small; font-family: arial,helvetica,sans-serif; }
-->
</STYLE>
<title>ploticus: pl(1)</title>
<body bgcolor=D0D0EE vlink=0000FF>
<br>
<br>
<center>
<table cellpadding=2 bgcolor=FFFFFF width=550><tr>
<td>
  <table cellpadding=2 width=550><tr>
  <td><br><h2>pl(1)</h2></td>
  <td align=right>
  <small>
  <a href="../doc/welcome.html"><img src="../doc/ploticus.gif" border=0></a><br>
  Version 2.40 Jan'08
  <td></tr></table>
</td></tr>
<td>
<br>
<br>
.>>

.TH pl(1) PL "15-JAN-2008   PL ploticus.sourceforge.net"

.SH Name
pl

.SH Synopsis
\fCpl  -prefab \fIprefabname\fC   \fIparm=value\fC ..  [\fI-options\fR]
.br
\0.. OR ..
.br
\fCpl \fIscriptfile\fC  [\fI-options\fC]  

.ig >>
<br><br><br>
.>>
.ig >>
<table cellpadding=2 border=0 ><tr bgcolor=FFFFEC><td colspan=2>
<b>Quick links for finding things on this page:</b>
</td></tr><tr bgcolor=FFFFEC><td valign=top>
<ul>
  <li> <a href="#description">Description</a>
  <li> <a href="#findexamples">Where to find examples</a>
  <li> <a href="#options">Command line arguments</a>
  <br> &nbsp; &nbsp; <a href="#basic">Basic options</a>
  <br> &nbsp; &nbsp; <a href="#varvalue">Setting ploticus variables</a>
  <br> &nbsp; &nbsp; <a href="#resultformat">Result format options</a>
  <br> &nbsp; &nbsp; <a href="#clickmap">Clickmap and mouseover</a>
  <br> &nbsp; &nbsp; <a href="#resultsize">Result sizing options</a>
  <br> &nbsp; &nbsp; <a href="#graphic">Graphics environment options</a>
  <br> &nbsp; &nbsp; <a href="#capacity">Capacity setting options</a>
  <br> &nbsp; &nbsp; <a href="#svg">SVG / XML options</a>
  <br> &nbsp; &nbsp; <a href="#interact">Interactive / workstation options</a>
  <br> &nbsp; &nbsp; <a href="#paper">Paper orientation options</a>
  <br> &nbsp; &nbsp; <a href="#debug">Development & debugging options</a>
</td><td valign=top>
  <li> <a href="#outfile">Naming of the result file(s)</a>
  <li> <a href="#examples">Usage examples</a>
  <li> <a href="#environment">Environment</a>
  <li> <a href="#bugs">Bugs</a>
  <li> <a href="#author">Author, Copyright, Licensing</a>
  <li> <a href="#seealso">See also</a>
</td></tr></table>
.>>

.ig >>
<a name=description></a>
.>>
.ig >>
<br><br><br>
.>>
.SH Description
\fBpl\fR is the primary component of the 
.ig >>
<a href="http://ploticus.sourceforge.net">
.>>
\0'ploticus' data display package
.ig >>
</a>
.>>
.LP
\fBpl\fR is a program that produces plots and charts from data, and 
produces results that can be viewed on web pages, paper, slides,
or interactively on the screen.  Standard types of plots may be
done using
.ig >>
<a href="prefabs.html">
.>>
\0prefab plot templates
.ig >>
</a>
.>>
, or a user-developed 
.ig >>
<a href="scripts.html">
.>>
\0script file
.ig >>
</a>
.>>
may be supplied for greater flexibility and customization.
\fBpl\fR may be executed from the command line or 
.ig >>
<a href="cgi.html">
.>>
\0as a CGI program.
.ig >>
</a>
.>>
.LP
For complete online docs and downloads see 
.ig >>
<a href="http://ploticus.sourceforge.net">
.>>
\0http://ploticus.sourceforge.net
.ig >>
</a>
.>>

.ig >>
<br><br><br>
.>>

.ig >>
<a name=findexamples></a>
.>>

.SH Where to find examples
See the various
.ig >>
<a href="prefabs.html">
.>>
\0prefab examples
.ig >>
</a>
.>>
\0.
A large number of 
.ig >>
<a href="../gallery/index.html">
.>>
\0script examples
.ig >>
</a>
.>>
are also available.
Some 
.ig >>
<a href="#examples">
.>>
\0usage examples
.ig >>
</a>
.>>
are also shown below.


.ig >>
<br><br><br>
.>>
.ig >>
<a name=options></a>
<br>
.>>

.SH Command line arguments

.LP
Command line arguments may generally be given in any order.
If there are arguments that you want to always have in effect, you can invoke them from a
.ig >>
<a href="config.html">
.>>
\0config file.
.ig >>
</a>
.>>
Many settings can also be made dynamically from scripts via
.ig >>
<a href="settings.html">
.>>
\0proc settings
.ig >>
</a>
.>>
or
.ig >>
<a href="page.html">
.>>
\0proc page.
.ig >>
</a>
.>>
Processing occurs in this order: first the config file is read; then command line args are processed (left to right); 
then 
.ig >>
<a href="page.html">
.>>
\0proc page
.ig >>
</a>
.>>
and/or 
.ig >>
<a href="settings.html">
.>>
\0proc settings.
.ig >>
</a>
.>>
Later settings override earlier ones.


.ig >>
<a name=basic></a>
.>>
.ig >>
<br><br>
.>>

.SH Basic command line options
.LP
\fB-prefab\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIprefabname\fR
.IP \0
Produce a plot using a 
.ig >>
<a href="prefabs.html">
.>>
\0prefab
.ig >>
</a>
.>>
plot template.
\fIprefabname\fR identifies the template, eg.
\fCcron\fR or \fCvbars\fR.
Necessary parameters are supplied on the command line using the form \fIparm=value\fR.

.ig >>
<br><br>
.>>
.LP
\fIscriptfile\fR  
.br
\fB-f \fIscriptfile\fR
.IP \0
names a
.ig >>
<a href="scripts.html">
.>>
\0script file
.ig >>
</a>
.>>
that will be interpreted to produce results.
Alternatively, \fB-stdin\fR may be used to indicate
that script will be available on standard input.

.ig >>
<br><br>
.>>

.ig >>
<a name=varvalue></a>
.>>
.LP
\fIvariable\fC=\fIvalue\fR
.IP \0
Declares the named
\fIvariable\fR and sets it to the given \fIvalue\fR.  
This is a convenient way to pass information
to prefabs and scripts.  Variable names are case-sensitive.
.br
Example: \fCCUTDATE=10-31-98\fR
.br
sets the variable \fCCUTDATE\fR to \fC10-31-98\fR.

.ig >>
<br><br>
.>>

.LP
\fB-o 
.ig >>
&nbsp; &nbsp;
.>>
\fIoutfile\fC | stdout\fR
.IP \0
Specify a filename where the result will be written.
No processing is applied to this name.. so the ending should be appropriate for
the selected output format, eg. use \fC.png\fR for PNG files.
If \fC-o stdout\fR is used,
result will be sent to standard output.  
If \fB-o\fR is not specified, a 
.ig >>
<a href="#outfile">
.>>
\0default output filename
.ig >>
</a>
.>>
will be used.  
.br
Example: \fC-o fp001.png\fR

.ig >>
<br><br>
.>>

.LP
\fB-dir 
.ig >>
&nbsp; &nbsp;
.>>
\fIdirname\fR
.IP \0
Set ploticus' working directory to \fIdirname\fR.
If used, this argument should be specified leftmost on the command line, 
since it affects evaluation of other args.


.ig >>
<a name=resultformat></a>
.>>

.ig >>
<br><br><br>
.>>

.SH Result format options
.LP
(Availability depends on your ploticus configuration/build)
.LP
\fB-png \fR 
.ig >>
&nbsp; &nbsp; &nbsp;
.>>
PNG image

.LP
\fB-gif \fR
.ig >>
&nbsp; &nbsp; &nbsp; &nbsp;
.>>
pseudo-GIF image

.LP
\fB-jpeg\fR
.ig >>
&nbsp; &nbsp; 
.>>
JPEG image 

.LP
\fB-svg\fR or \fB-svgz\fR
.ig >>
&nbsp; &nbsp; &nbsp;
.>>
.ig >>
<a href="svg.html">
.>>
\0SVG
.ig >>
</a>
.>>
graphic.  
See also SVG / XML options below.

.LP
\fB-swf\fR
.ig >>
&nbsp; &nbsp; &nbsp;
.>>
.ig >>
<a href="swf.html">
.>>
\0SWF
.ig >>
</a>
.>>
(flash) result.

.LP
\fB-wbmp\fR
.ig >>
&nbsp; 
.>>
WBMP image


.LP
\fB-eps \fR
.ig >>
&nbsp; &nbsp; &nbsp;
.>>
EPS (encapsulated PostScript) 

.LP
\fB-ps \fR
.ig >>
&nbsp; &nbsp; &nbsp; &nbsp; 
.>>
paginated PostScript to stdout

.LP
\fB-x11 \fR
.ig >>
&nbsp; &nbsp; &nbsp; &nbsp; 
.>>
display on X11 screen

.LP
\fB-drawdump\fR
\fIfilename\fR
.ig >>
&nbsp; &nbsp;
.>>
produce no visible graphic; save a generic representation of the graphic result to a file.
By using \fC-drawdump\fR and \fC-drawdumpa\fR you can easily 
.ig >>
<a href="prefabs.html#drawdump">
.>>
\0overlay or combine results from separate pl runs.
.ig >>
</a>
.>>
The drawdump file can be rendered later in any desired format, using this command:
\fCpl -prefab draw dumpfile=\fIfilename\fR 
.ig >>
&nbsp; &nbsp;
.>>
or by using
.ig >>
<a href="drawcommands.html">
.>>
\0proc drawcommands.
.ig >>
</a>
.>>
Drawdump capability is available in all builds. (2.30+)
.LP
\fB-drawdumpa\fR
\fIfilename\fR
.ig >>
&nbsp; &nbsp;
.>>
same as \fC-drawdump\fR but result is \fBappended\fR to file.



.ig >>
<a name=clickmap></a>
.>>

.ig >>
<br><br><br>
.>>
.SH Clickable image maps and mouseover options

.LP
\fB-csmap\fR
.IP \0
produce a 
.ig >>
<a href="clickmap.html">
.>>
\0client-side clickable imagemap
.ig >>
</a>
.>>
to accompany a png, gif, or jpeg.
These can be used for hyperlinks, and also for providing
pop-up text labels that appear when the mouse passes over a region.
By default, client-side map content is written to stdout.

.ig >>
<br><br>
.>>

.LP
\fB-csmapdemo\fR
.IP \0
Same as \fB-csmap\fR but all mapped regions are shown outlined in green,
and a complete HTML chunk is produced which involves the output image name.

.ig >>
<br><br>
.>>

.LP
\fB-mapfile\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIfilename\fC | stdout | stderr\fR
.IP \0
explicitly name the output file containing the map info.  The name may also be set in 
.ig >>
<a href="page.html">
.>>
\0proc page.
.ig >>
</a>
.>>
If a name is not specified, client-side image map info will be written to stdout;
For SVG this parameter is not needed, since image map info is embedded in the SVG file.

.ig >>
<br><br>
.>>
.LP
\fB-map\fR
.IP \0
produce a 
.ig >>
<a href="clickmap.html">
.>>
\0server-side clickable imagemap file
.ig >>
</a>
.>>
to accompany a png, gif, jpeg, or SVG.   
\fBNote: server-side maps are deprecated.\fR

.ig >>
<a name=resultsize></a>
.>>
.ig >>
<br><br><br>
.>>

.SH Result sizing options
.LP
\fB-scale\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIsx\fR[,\fIsy\fR]
.IP \0
Scale the final result.
If one value is given, the result is scaled by this amount in both x and y.
If two values are given, scaling in x and scaling in y may be done independently.
A scale value of less than 1.0 reduces the
size; an scale value of greater than 1.0 enlarges.  
Scaling is done relative to the origin (0,0) which is at the lower left.
.br
Example: \fC-scale 0.7\fR

.ig >>
<br><br>
.>>

.LP
\fB-pagesize\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIwidth,height\fR
.IP \0
Sets the pre-crop size of the result image for GIF/PNG/JPEG, or sets the display window size when drawing to X11.  
On other output devices this option does nothing.
\fIwidth\fR and \fIheight\fR are in 
.ig >>
<a href="attributetypes.html#positionunits">
.>>
\0absolute units.
.ig >>
</a>
.>>
0,0 is the lower left corner.
If -pagesize is not specified, the default size will be 8" x 8".
Size is set before any drawing takes place and is unaffected by the -scale option.
.IP
When rendering PNG/GIF/JPEG images, this option determines amount of internal memory allocation for accommodating the image.
The result can never be bigger than this size, and any drawing outsize the bounds will not be visible.
\fBTo create PNG/GIF/JPEG images larger than 8" x 8", this option MUST be specified to set a bigger size.\fR
Cropping options (below) can be used along with -pagesize as long as they result in a smaller rectangle
than the pagesize; they take effect after all drawing has been completed.
.IP
\fB-pagesize\fR has no effect with EPS or paginated PostScript results (the PostScript
BoundingBox will be determined by the extent of the graphic).
.IP
Example: \fC-pagesize 7,3\fR

.ig >>
<br><br>
.>>

.LP
\fB-tightcrop\fR
.IP \0
For image or EPS output, crop the result tightly to the extent of
the design.  Normally a small margin is allowed on all four sides.
This option sometimes crops a bit too tight; if so try \fB-croprel\fR.

.ig >>
<br><br>
.>>

.LP
\fB-crop\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIx1,y1,x2,y2\fR
.IP \0
Crop image or EPS result to the box specified by
\fIx1,y1\fR and \fIx2,y2\fR, in 
.ig >>
<a href="attributetypes.html#positionunits">
.>>
\0absolute units.
.ig >>
</a>
.>>

Note that there may be no spaces in the coordinates specification.
Cropping takes place after design is rendered and does not
affect coordinate locations.
.br
Example: \fC-crop 1.2,0.8,4.4,5.2\fR

.ig >>
<br><br>
.>>

.LP
\fB-croprel\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIleft,bottom,right,top\fR
.IP \0
Crop image or EPS result tightly to the extent of the design (like \fB-tightcrop\fR),
but then adjust the cropping outward or inward on one or more sides.
\fIleft\fR is the amount to adjust the left side, in 
.ig >>
<a href="attributetypes.html#positionunits">
.>>
\0absolute units.
.ig >>
</a>
.>>
Similarly for \fIbottom\fR, \fIright\fR, and \fItop\fR.
Positive values always adjust outward from center; negative values adjust inward (tighter).
There may be no spaces in the \fIleft,bottom,right,top\fR specification.
Cropping takes place after design is rendered and does not
affect coordinate locations.
.br
Example: \fC-croprel 0,-0.1,0,0.1\fR

.ig >>
<br><br>
.>>

.LP
\fB-pixsize\fR
.ig >>
&nbsp; &nbsp;
.>>
\fIwidth,height\fR
.IP \0
If specified, result PNG/GIF/JPG image will be created at exactly this width and height in pixels.
Does not interact with scaling or cropping... user is responsible for ensuring that content fits appropriately into the specified size.
User is also responsible for setting -pagesize appropriately for larger images.
New in 2.40

.ig >>
<a name=graphic></a>
.>>

.ig >>
<br><br><br>
.>>


.SH Graphics environment options
.LP
\fB-font\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIfont\fR 
.IP \0
sets the overall font to \fIfont\fR.  See
.ig >>
<a href="fonts.html">
.>>
\0fonts
.ig >>
</a>
.>>
for more info.

.ig >>
<br><br>
.>>

.LP
\fB-textsize\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIpointsize\fR 
.IP \0
sets the overall default textsize to \fIpointsize\fR.
All embedded size specifications will be rendered relative to this.

.ig >>
<br><br><br>
.>>

.LP
\fB-linewidth\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIw\fR 
.IP \0
sets the overall default linewidth to \fIw\fR.
All embedded line width specifications will be rendered relative to this.
See linedetails(pli) for more on line width.

.ig >>
<br><br>
.>>
.LP
\fB-color\fR 
.ig >>
&nbsp; &nbsp;
.>>
.ig >>
<a href="color.html">
.>>
\0color
.ig >>
</a>
.>>
.IP \0
sets the overall default text and line drawing color to \fIcolor\fR.

.ig >>
<br><br>
.>>

.LP
\fB-backcolor\fR 
.ig >>
&nbsp; &nbsp;
.>>
.ig >>
<a href="color.html">
.>>
\0color
.ig >>
</a>
.>>
.IP \0
sets the background color to \fIcolor\fR.

.ig >>
<br><br>
.>>

.ig >>
<a name=cm></a>
.>>
.LP
\fB-cm\fR
.ig >>
&nbsp; &nbsp;
.>>
.IP \0
Use centimeters as your absolute units, instead of inches.
On the command line this must appear to the left of any arguments dealing with
absolute unit values, such as -pagesize.
Centimeter absolute units can also be set via
.ig >>
<a href="../doc/settings.html">
.>>
\0proc settings.
.ig >>
</a>
.>>
If cm will always be the desired
absolute units, the preferred way to achieve this is by using 
\fCunits: cm\fR in a 
.ig >>
<a href="../doc/config.html">
.>>
\0ploticus config file.
.ig >>
</a>
.>>

.ig >>
<br><br>
.>>

.LP
\fB-inches\fR
.IP \0
Use inches as your absolute units.  This is the default.

.ig >>
<br><br>
.>>

.LP
\fB-outlabel\fR  
.ig >>
&nbsp; &nbsp;
.>>
\fIlabel\fR
.IP \0
Set the label or title for the output.  For X11 this sets the window title;
for PostScript and SVG it sets the \fC%%Title\fR attribute.

.ig >>
<a name=capacity></a>
.>>

.ig >>
<br><br><br>
.>>
.SH Capacity setting options
These options (new with version 2.10) allow capacities to be raised for accomodation 
of very large data sets, or lowered to minimize memory usage.
The defaults in this section are defined in \fCpl.h\fR.
.LP
\fB-maxrows 
.ig >>
&nbsp; &nbsp;
.>>
\fInrows\fR
.IP \0
Set the capacity for data rows to \fInrows\fR.  Default \fInrows\fR is 10,000.
Ploticus will allocate one pointer for each row.

.ig >>
<br><br>
.>>

.LP
\fB-maxfields 
.ig >>
&nbsp; &nbsp;
.>>
\fInfields\fR
.IP \0
Set the capacity for data fields to \fInfields\fR.  Default \fInfields\fR is 200,000.
Ploticus will allocate one pointer for each field.

.ig >>
<br><br>
.>>

.LP
\fB-maxproclines  
.ig >>
&nbsp; &nbsp;
.>>
\fInlines\fR
.IP \0
Set the capacity for script lines for active procs to \fInlines\fR.
Default \fInlines\fR is 5000.
Active procs are the current proc, all \fC#save\fRd procs, and
all proc getdata procs that contain embedded \fCdata\fR.
Ploticus will allocate one pointer for each line in each active proc.

.ig >>
<br><br>
.>>

.LP
\fB-maxvector  
.ig >>
&nbsp; &nbsp;
.>>
\fIncells\fR
.IP \0
Set the capacity for the data plotting vector to \fIncells\fR.
Default \fIncells\fR is 100,000.
The data plotting vector is an array which holds plottable
values for situations where the values must be sorted or pre-screened for
bad values.  Ploticus will allocate one \fCdouble\fR value for each cell.

.ig >>
<br><br>
.>>
.LP
\fB-maxdrawpoints
.ig >>
&nbsp; &nbsp;
.>>
\fIn\fR
.IP \0
Use this if you need to render a polygon having more than 500 points in PNG/GIF/JPEG, X11, or SWF, or
any continuous line having more than 500 points in SWF.

.ig >>
<br><br>
.>>
.LP
Note: raising the maximum number of categories may be done using
.ig >>
<a href="categories.html">
.>>
\0proc categories
.ig >>
</a>
.>>
from within the script.

.ig >>
<br><br>
.>>
.LP
\fB-cpulimit\fR
#Include nbsp2
\fIs\fR
.IP \0
Set unix resource limit on cpu time to \fIs\fR seconds.  Default is 30 seconds.  
New in 2.40


.ig >>
<br><br><br>
.>>
.ig >>
<a name=svg></a>
.>>

.SH SVG / XML options:

.LP
\fB-svg_tagparms\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIstring\fR
.IP \0
This allows arbitrary \fCtext\fR to be inserted into the opening \fC<svg>\fR tag.
.br
Example: \fC-svg_tagparms 'height="10cm" width="15cm"'\fR

.ig >>
<br><br>
.>>

.LP
\fB-omit_xml_declaration\fR  
.IP \0
By default the first line of the SVG result will be the XML declaration \fC<?xml .. >\fR .
Use this option to suppress the XML declaration line if
the SVG result is to be embedded into a larger XML document.

.ig >>
<br><br>
.>>

.LP
\fB-xml_encoding\fR  
.ig >>
&nbsp; &nbsp;
.>>
\fImethod\fR
.IP \0
Set the XML character encoding method.  This encoding will be indicated in the XML declaration line.
The default is \fCiso-8859-1\fR which provides Latin and Western European character sets.
For Unicode fonts this should be set to \fCutf-8\fR (for more discussion see the Unicode section in
.ig >>
<a href="fonts.html">
.>>
\0fonts
.ig >>
</a>
.>>
).

.ig >>
<br><br>
.>>

.LP
\fB-tag\fR 
.IP \0
Causes a suitable
HTML \fC<EMBED>\fR tag to be written to standard output.  

.ig >>
<br><br>
.>>

.LP
\fB-zlevel \fIn\fR 
.IP \0
This may be used to set the compression level to \fIn\fR for SVGZ output
(0 - 9 where 9 is highest level of compression and the default).


.ig >>
<a name=interact></a>
.>>

.ig >>
<br><br><br>
.>>
.SH Interactive (workstation) use options 
.LP
\fB-winloc\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIx,y\fR
.IP \0
Control where on the screen the upper-left corner of the X11 display
window will be placed.  \fIx\fR and \fIy\fR are in pixels.
Example: \fC-winloc 200 0\fR

.ig >>
<br><br>
.>>

.LP
\fB-v\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIcommand\fR
.br
\fB-viewer\fR 
.ig >>
&nbsp; &nbsp;
.>>
\fIcommand\fR
.IP \0

graphics/pl240src/src/pl.c
/* ======================================================= *
 * Copyright 1998-2008 Stephen C. Grubb                    *
 * http://ploticus.sourceforge.net                         *
 * Covered by GPL; see the file ./Copyright for details.   *
 * ======================================================= */

/* PL - ploticus main module */

#include "pl.h"
#include <unistd.h>
#ifdef WIN32
#include <fcntl.h>  /* for _O_BINARY */
#endif

extern int PLGG_setimfmt();


/* ============================================= */
int
PL_version_msg( longmsg )
int longmsg;
{
char outputformats[80];
PL_devstring( outputformats );


#ifdef WIN32
fprintf( PLS.diagfp, "ploticus %s (win32).", PLVERSION );
#else
fprintf( PLS.diagfp, "ploticus %s (unix).", PLVERSION );
#endif
fprintf( PLS.diagfp, "  %s\n", outputformats );
fprintf( PLS.diagfp, "Copyright 1998-2008 Steve Grubb, http://ploticus.sourceforge.net\n\n" );

if( longmsg ) fprintf( PLS.diagfp, "Please see the Copyright file or web site for additional credits and information.\n\
\n\
This program is free software; you can redistribute it and/or modify it\n\
under the terms of the GNU General Public License as published by the\n\
Free Software Foundation; either version 2 of the License, or (at your\n\
option) any later version.\n\
\n\
This program is distributed in the hope that it will be useful, but \n\
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY \n\
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License\n\
for more details.\n" );
exit( 1 );
}

/* ============================================= */
/* MAIN */
/* ============================================= */
int
main( argc, argv )
int argc;
char **argv;
{
int argi, use_stdin, stat, found, valused, stoparg, ci, cii;
char buf[256];
char scriptfile[MAXPATH];
char prefabname[80];
char *arg, *nextarg, *getenv();
char *outnamebase;





/* check to see if we are operating as a direct CGI program.. */
PLS.cgiargs = NULL;
if( argc == 1 ) {
	PLS.cgiargs = getenv( "QUERY_STRING" );
	if( getenv( "SCRIPT_FILENAME" )==NULL ) {  /* some web servers don't set SCRIPT_FILENAME.. scg 8/27/02 */
		sprintf( buf, "SCRIPT_FILENAME=%s", argv[0] );
		putenv( buf );
		}
	}


/* set hard defaults and process config file if any.. */
stat = PL_do_preliminaries();
if( stat ) exit( 1 );
use_stdin = 0;


/* if in direct cgi mode, initialize and set some useful defaults.. */
if( PLS.cgiargs != NULL ) {
	/* begin parsing QUERY_STRING.. */
	stoparg = 80; 
	ci = 0;
	
	strcpy( PLS.outfile, "stdout" );
#ifdef WIN32
        /* must set stdin and stdout to binary mode */
        _setmode( _fileno( stdin ), _O_BINARY );
        _setmode( _fileno( stdout ), _O_BINARY );
#endif /* WIN32 */

	PLS.device = 's';
#ifndef NOGD
	PLS.device = 'g';
	/* try to be smart about picking default output format.. */
	if( devavail( "gif" )) PLGG_setimfmt( "gif" );
	else if( devavail( "jpeg" )) PLGG_setimfmt( "jpeg" );
	else if( devavail( "png" )) PLGG_setimfmt( "png" );
#endif
	}

else stoparg = argc;

strcpy( scriptfile, "" );
strcpy( prefabname, "" );

/* process command line arguments.. (if direct CGI, arguments are parsed from URL).. */
for( argi = 1; argi < stoparg; argi++ ) {

	if( PLS.cgiargs != NULL ) {
		/* parse next 2 args from QUERY_STRING.. */
		GL_getcgiarg( PL_bigbuf, PLS.cgiargs, &ci, 252 ); 
		if( PL_bigbuf[0] == '\0' ) break;
		arg = PL_bigbuf;
		cii = ci;
		GL_getcgiarg( &PL_bigbuf[500], PLS.cgiargs, &cii, 252 );  /* share PL_bigbuf */
		nextarg = &PL_bigbuf[500]; 
		}
	else	{
		arg = argv[ argi ];
		if( argi+1 < argc ) nextarg = argv[argi+1];
		else nextarg = "";
		}

	if( strcmp( arg, "-stdin" )==0 && PLS.cgiargs == NULL ) use_stdin = 1;


#ifdef HOLD
	else if( strcmp( arg, "-png" )==0 && !devavail( "png" ) && PLS.cgiargs == NULL ) { 
		char *p[100];
		if( strlen( argv[0] ) >= 5 ) {
			if( strcmp( &argv[0][strlen(argv[0])-5], "plpng" )==0 ) {
				Eerr( 5279, "png not available in plpng", "" );
				PL_version_msg( 0 ); exit(1);
				}
			}
		p[0] = "plpng";
		for( i = 1; i < argc; i++ ) p[i] = argv[i];
		p[argc] = NULL;
		execvp( "plpng", argv );
		fprintf( PLS.errfp, "PNG not supported in this build (plpng not found).\n" );
                PL_version_msg( 0 ); exit(1);
		}
#endif

	else if( strcmp( arg, "-f" )==0 ) {
		if( strlen( nextarg ) > MAXPATH-10 ) { /* allow extra for output file suffix add */
			fprintf( PLS.errfp, "pl: script file name too long" );
			PL_version_msg( 0 ); exit( 1 );
			}
		strcpy( scriptfile, nextarg );
		argi++;
		}

	else if( strcmp( arg, "-prefab" )==0 ) {
		if( PLS.cgiargs != NULL ) {
			Eerr( 4916, "-prefab not available in direct cgi mode", "" ); PL_version_msg( 0 ); exit( 1 );
			}
		if( PLS.prefabsdir == NULL ) {
			Eerr( 4899, "PLOTICUS_PREFABS environment var not found (pathname of dir where prefab files reside)", "" );
			PL_version_msg( 0 ); exit( 1 );
			}
		sprintf( prefabname, "%s.pl", nextarg );
		sprintf( scriptfile, "%s%c%s", PLS.prefabsdir, PATH_SLASH, prefabname );
		if( PLS.debug ) fprintf( PLS.diagfp, "Prefabs dir is: %s\nScript file is %s\n", PLS.prefabsdir, scriptfile );
		argi++;
		}

	else if( GL_smember( arg, "-? -help -ver -version" ) ) { PL_version_msg( 1 ); exit(0); }

	else 	{
		stat = PL_process_arg( arg, nextarg, &valused, &found );
		if( stat != 0 ) exit( 1 );
		if( !found && arg[0] == '-' ) Eerr( 4892, "warning, unrecognized argument", arg );
		else if( !found && scriptfile[0] == '\0' ) {
			if( strlen( arg ) > MAXPATH-10 ) { /* allow extra for output file suffix add */
				fprintf( PLS.errfp, "pl: script file name too long" );
				PL_version_msg( 0 ); exit( 1 );
				}
			strcpy( scriptfile, arg  );  
			}
		argi += valused;
		if( PLS.cgiargs != NULL && valused ) ci = cii; /* jump ahead */
		}
	}



/* CGI header stuff.. */
if( PLS.cgiargs != NULL ) {
	char imagetype[20];
	strcpy( PLS.outfile, "stdout" );
	/* check for loopy script file names.. */
	if( scriptfile[0] == '/' && prefabname[0] == '\0' ) {   /* changed scg 2/6/02 */
		Eerr( 2740, "cgi mode: script file name may not begin with '/'", scriptfile );
		PL_version_msg( 0 ); exit(1);
		}
	if( GL_slmember( scriptfile, "*..* .*" ) ) {
		Eerr( 2740, "cgi mode: script file name may not begin with '.' or contain '..'", scriptfile );
		PL_version_msg( 0 ); exit(1);
		}
	/* output the HTTP content-type header.. */
	devnamemap( &(PLS.device), imagetype, 2 );
	if( PLS.device == 's' ) printf( "Content-type: image/%s-xml\n\n", imagetype );
	else printf( "Content-type: image/%s\n\n", imagetype );

	/* be sure clickmap is off - incompatible with direct cgi mode.. */
	PLS.clickmap = 0;
	}



/* if script coming from stdin, copy stdin to a tmp file.. */
if( use_stdin ) {
	FILE *tfp;
	sprintf( scriptfile, "%s_I", PLS.tmpname );
	tfp = fopen( scriptfile, "w" ); /* temp file, unlinked below */
	if( tfp == NULL ) { Eerr( 102, "Cannot open tmp file for stdin script\n", scriptfile ); PL_version_msg( 0 ); exit(1); }
	while( fgets( PL_bigbuf, MAXBIGBUF-1, stdin ) != NULL ) fprintf( tfp, "%s", PL_bigbuf ); /* was 255 scg 5/20/03 */
	fclose( tfp );
	}
	

if( scriptfile[0] == '\0' ) {
	fprintf( PLS.diagfp, "usage: pl scriptfile [options] ...or...  pl -prefab prefabname [options]\n" );
	PL_version_msg( 0 );
	}



/* DEVICE variable.. */

if( TDH_getvar( "DEVICE", buf ) != 0 ) { /* DEVICE not given on command line, set DEVICE */
	stat = devnamemap( &(PLS.device), buf, 2 );
	if( stat != 0 ) { PL_version_msg( 0 ); exit( 1 ); }
	TDH_setvar( "DEVICE", buf );
	}
else	{ /* DEVICE given on command line, set PLS.device from DEVICE */
	TDH_getvar( "DEVICE", buf );
	stat = devnamemap( &(PLS.device), buf, 1 );
	if( stat != 0 ) { PL_version_msg( 0 ); exit( 1 ); }
	}
if( PLS.debug ) fprintf( PLS.diagfp, "Device code is %c\n", PLS.device );



/* build output file names.. */

if( prefabname[0] != '\0' ) outnamebase = prefabname;
else outnamebase = scriptfile;

if( PLS.outfile[0] == '\0' && GL_member( PLS.device, "egsf" ) ) makeoutfilename( outnamebase, PLS.outfile, PLS.device, 1);


/* if a viewcommand is specified and an outfile has not been and the device 
   is paginated postscript, we need to set the output file to out.ps */
if( PLS.viewer[0] != '\0' && PLS.outfile[0] == '\0' && GL_member( PLS.device, "cp")) strcpy( PLS.outfile, "out.ps" );

if( PLS.outfile[0] != '\0' && 
    strcmp( PLS.outfile, "-" )!= 0 ) {
	if( PLS.debug ) fprintf( PLS.diagfp, "Setting output file name to %s\n", PLS.outfile );
	Esetoutfilename( PLS.outfile );
	}


PL_begin(); /* various other initializations that must be done after config & args processing.. */



/* execute the script file to produce the plot.. */
if( PLS.debug ) fprintf( PLS.diagfp, "Script file is: %s\n", scriptfile );
stat = PL_exec_scriptfile( scriptfile );
if( stat != 0 ) { PL_version_msg( 0 ); exit( 1 ); }


/* finish up (x11: button, etc.) */
if( PLS.eready && PLS.device == 'x' ) PL_do_x_button( "End." );

if( PLS.eready ) Eendoffile();
if( use_stdin ) unlink( scriptfile );

if( PLS.viewer[0] != '\0' && PLS.device != 'x' && PLS.cgiargs == NULL ) {
	int len;
	strcpy( buf, PLS.viewer );
	len = strlen( buf );
	strcpy( &buf[len++], " " );
	Egetoutfilename( &buf[ len ] );
	if( strnicmp( &buf[ len ], "stdout", 6 )==0 ) fprintf( PLS.diagfp, "Cannot use -o stdout with -viewer\n" );
	else if( PLS.noshell ) fprintf( PLS.diagfp, "-noshell prohibits -viewer" );
	else 	{ 
		fprintf( PLS.diagfp, "Executing '%s' to view results..\n", buf ); 
		system( buf ); 
		}
	}

PL_free();
exit( 0 );
}

/* ======================================================= *
 * Copyright 1998-2008 Stephen C. Grubb                    *
 * http://ploticus.sourceforge.net                         *
 * Covered by GPL; see the file ./Copyright for details.   *
 * ======================================================= */
